AmplifyPlugins/Auth/Sources/AWSCognitoAuthPlugin/Task/AWSAuthConfirmSignInTask.swift

```diff
@@ -49,17 +49,32 @@ class AWSAuthConfirmSignInTask: AuthConfirmSignInTask, DefaultLogger {
             AuthPluginErrorConstants.invalidStateError, nil)
 
         guard case .configured(let authNState, _) = await authStateMachine.currentState,
-              case .signingIn(let signInState) = authNState,
-              case .resolvingChallenge(let challengeState, _, _) = signInState else {
+              case .signingIn(let signInState) = authNState else {
             throw invalidStateError
         }
 
-        switch challengeState {
-        case .waitingForAnswer, .error:
-            log.verbose("Sending confirm signIn event: \(challengeState)")
-            await sendConfirmSignInEvent()
-        default:
-            throw invalidStateError
+        if case .resolvingChallenge(let challengeState, let challengeType, _) = signInState {
+
+            // Validate if request valid MFA selection
+            if case .selectMFAType = challengeType {
+                try validateRequestForMFASelection()
+            }
+
+            switch challengeState {
+            case .waitingForAnswer, .error:
+                log.verbose("Sending confirm signIn event: \(challengeState)")
+                await sendConfirmSignInEvent()
+            default:
+                throw invalidStateError
+            }
+        } else if case .resolvingTOTPSetup(let resolvingSetupTokenState, _) = signInState {
+            switch resolvingSetupTokenState {
+            case .waitingForAnswer, .error:
```

Comment:
Retry on TOTP failure is allowed?

Reply:
Yes. I have also added integration tests for it.